core/src/replay_stage.rs

```diff
@@ -644,72 +666,145 @@ impl ReplayStage {
+        //highest weight, lowest slot first
+        votable.sort_by_key(|b| (b.1.weight, 0i64 - b.1.slot as i64));
 
+        votable.iter().for_each(|(_, stats)| {
+            let mut parents: Vec<_> = if let Some(set) = ancestors.get(&stats.slot) {
+                set.iter().collect()
+            } else {
+                vec![]
+            };
+            parents.sort();
+            debug!("{}: {:?} {:?}", stats.slot, stats, parents,);
+        });
         trace!("votable_banks {}", votable.len());
-        if !votable.is_empty() {
-            let weights: Vec<u128> = votable.iter().map(|x| x.0).collect();
-            info!(
-                "@{:?} tower duration: {:?} len: {} weights: {:?}",
-                timing::timestamp(),
-                ms,
-                votable.len(),
-                weights
-            );
-        }
+        let rv = Self::pick_best_fork(ancestors, &votable);
+        let ms = timing::duration_as_ms(&tower_start.elapsed());
+        let weights: Vec<(u128, u64, u64)> = votable
+            .iter()
+            .map(|x| (x.1.weight, x.1.slot, x.1.block_height))
+            .collect();
+        debug!(
+            "@{:?} tower duration: {:?} len: {}/{} weights: {:?} voting: {}",
+            timing::timestamp(),
+            ms,
+            votable.len(),
+            stats.iter().filter(|s| !s.has_voted).count(),
+            weights,
+            rv.0.is_some()
+        );
         inc_new_counter_info!("replay_stage-tower_duration", ms as usize);
+        rv
+    }
 
-        votable
+    fn pick_best_fork(
+        ancestors: &HashMap<u64, HashSet<u64>>,
+        best_banks: &[(&Arc<Bank>, &ForkStats)],
+    ) -> (Option<(Arc<Bank>, u64)>, Option<Arc<Bank>>) {
+        if best_banks.is_empty() {
+            return (None, None);
+        }
+        let mut rv = None;
+        let (best_bank, best_stats) = best_banks.last().unwrap();
+        debug!("best bank: {:?}", best_stats);
+        let mut by_slot: Vec<_> = best_banks.iter().collect();
+        by_slot.sort_by_key(|x| x.1.slot);
+        //look for the oldest ancestors of the best bank
+        if let Some(best_ancestors) = ancestors.get(&best_stats.slot) {
+            for (parent, parent_stats) in by_slot.iter() {
+                if parent_stats.is_locked_out {
+                    continue;
+                }
+                if !best_ancestors.contains(&parent_stats.slot) {
+                    continue;
+                }
+                debug!("best bank found ancestor: {}", parent_stats.slot);
+                inc_new_counter_info!("replay_stage-pick_best_fork-ancestor", 1);
+                rv = Some(((*parent).clone(), parent_stats.total_staked));
+            }
+        }
+        //look for the oldest child of the best bank
+        if rv.is_none() {
+            for (child, child_stats) in by_slot.iter().rev() {
+                if child_stats.is_locked_out {
+                    continue;
```

Comment:
Do you want to set the "heaviest" (second item in the tuple) to the latest child on this fork, even if the bank is locked out so we can later reset PoH recorder right to the oldest child?

Reply:
@carllin i think the right thing to do is to set poh to last vote or heaviest if no vote. Heaviest is always ahead of last vote when we can’t vote on it.

Reply:
It doesn't make sense to rest the Poh to something that doesn't attach to the last vote.

Reply:
Hmmm, what's the reasoning for:
1) If not locked out, voting for the oldest child of the best bank
2) If locked out, reset to the best bank, but don't reset to the oldest child of the best bank?

Reply:
The latter creates a fork, since that skips the best bank.